src/openrct2/world/Banner.cpp

```diff
@@ -379,11 +360,111 @@ void BannerElement::ResetAllowedEdges()
     AllowedEdges |= 0b00001111;
 }
 
+void UnlinkAllRideBanners()
+{
+    for (auto& banner : _banners)
+    {
+        if (!banner.IsNull())
+        {
+            banner.flags &= ~BANNER_FLAG_LINKED_TO_RIDE;
+            banner.ride_index = RIDE_ID_NULL;
+        }
+    }
+}
+
+void UnlinkAllBannersForRide(ride_id_t rideId)
+{
+    for (auto& banner : _banners)
+    {
+        if (!banner.IsNull() && (banner.flags & BANNER_FLAG_LINKED_TO_RIDE) && banner.ride_index == rideId)
+        {
+            banner.flags &= ~BANNER_FLAG_LINKED_TO_RIDE;
+            banner.ride_index = RIDE_ID_NULL;
+            banner.text = {};
+        }
+    }
+}
+
 Banner* GetBanner(BannerIndex id)
 {
-    if (id < std::size(_banners))
+    if (id < _banners.size())
+    {
+        auto banner = &_banners[id];
+        if (banner != nullptr && !banner->IsNull())
+        {
+            return banner;
+        }
+    }
+    return nullptr;
+}
+
+Banner* GetOrCreateBanner(BannerIndex id)
+{
+    if (id < MAX_BANNERS)
     {
+        if (id >= _banners.size())
+        {
+            _banners.resize(id + 1);
+        }
         return &_banners[id];
     }
     return nullptr;
 }
+
+Banner* CreateBanner()
+{
+    auto bannerIndex = BannerGetNewIndex();
+    auto banner = GetOrCreateBanner(bannerIndex);
+    if (banner != nullptr)
+    {
+        banner->id = bannerIndex;
+        banner->flags = 0;
+        banner->type = 0;
+        banner->text = {};
+        banner->colour = 2;
```

Comment:
Needs constants.

May be picked in a later PR (but please create an issue if you decide to do so).